### ISSUE TITLE
Redesign stack context structure

### ISSUE THREAD

**Issue by jessesquires** [jessesquires/JSQCoreDataKit#60]
After talking with Daniel Eggert, I found that nesting parent/child contexts is non-optimal.

Move from the existing stack architecture, to the following:

- Remove `mainContext` as a child context of the `backgroundContext`
- Connect both `backgroundContext` and `mainContext` to the coordinator
- Merge via notification APIs

--------------

Original design: 28b9e4e51bb755191db68a9ffbebe780480d81ce (what we want to go back to)
Refactor to parent-child: f0aa87c6e582c985238afac75cd2d50b72813d11

Basically need to revert stack changes in f0aa87c6e582c985238afac75cd2d50b72813d11 to go back to 28b9e4e51bb755191db68a9ffbebe780480d81ce

**Comment by jessesquires** [comment]
Update, let's do the following instead of a full overhaul:

Allow users to specify the current setup or the one described above by passing a configuration object to the stack.

**Comment by jessesquires** [comment]
Also:

http://stackoverflow.com/questions/29900193/core-data-parent-context-blocks-child/29918585#29918585

:smile:

**Comment by TimurBK** [comment]
I was actually quite surprised to find out that most of popular CoreData libraries use parent-child stack. On data set of hundreds it may add ~30% to fetch time(I got ~0.6s and by using main queue MOC directly connected to PSC I got 0.4s on iPhone 5). 
Stack in Daniel's answer(probably can be called "classic") is performing much better. And if it's not enough, two separate persistent store coordinators connected to single store may help.
